prometheus/prometheus.go

```diff
@@ -0,0 +1,92 @@
+// Copyright 2020 The go-commons Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package prometheus
+
+import (
+	"context"
+	"crypto/tls"
+	"fmt"
+	"net/http"
+	"time"
+
+	api "github.com/prometheus/client_golang/api"
+	apiv1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/common/model"
+)
+
+// Used to intercept and passed custom auth headers to prometheus client request
+func (bat authTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	if bat.username != "" {
+		req.SetBasicAuth(bat.username, bat.password)
+	}
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", bat.token))
+	return bat.Transport.RoundTrip(req)
+}
+
+// NewClient creates a prometheus struct instance with the given parameters
+func NewClient(url, token, username, password string, tlsSkipVerify bool) (*Prometheus, error) {
+	prometheus := Prometheus{
+		Endpoint: url,
+	}
+	cfg := api.Config{
+		Address: url,
+		RoundTripper: authTransport{
+			Transport: &http.Transport{Proxy: http.ProxyFromEnvironment, TLSClientConfig: &tls.Config{InsecureSkipVerify: tlsSkipVerify}},
+			token:     token,
+			username:  username,
+			password:  password,
+		},
+	}
+	c, err := api.NewClient(cfg)
+	if err != nil {
+		return &prometheus, err
+	}
+	prometheus.Api = apiv1.NewAPI(c)
+	// Verify Prometheus connection prior returning
+	if err := prometheus.verifyConnection(); err != nil {
+		return &prometheus, err
+	}
+	return &prometheus, nil
+}
+
+// Query prometheus query wrapper
+func (p *Prometheus) Query(query string, time time.Time) (model.Value, error) {
+	var v model.Value
+	v, _, err := p.Api.Query(context.TODO(), query, time)
+	if err != nil {
+		return v, err
+	}
+	return v, nil
+}
+
+// QueryRange prometheus queryRange wrapper
+func (p *Prometheus) QueryRange(query string, start, end time.Time, step time.Duration) (model.Value, error) {
+	var v model.Value
+	r := apiv1.Range{Start: start, End: end, Step: step}
+	v, _, err := p.Api.QueryRange(context.TODO(), query, r)
+	if err != nil {
+		return v, err
+	}
+	return v, nil
+}
```

Comment:
These 2 functions have been trimmed down from the original implementation from kube-burner, honestly, I think they don't add any value and they can be removed.

For example, a snippet of the kube-burner implementation from https://github.com/cloud-bulldozer/kube-burner/pull/317

```golang
v, err = p.Client.QueryRange(query, start, end, p.Step)
	if err != nil {
		log.Warnf("Error found with query %s: %s", query, err)
		continue
}
```
This can be easily replaced by:

```golang
v, err = p.Client.Api.QueryRange(query, start, end, p.Step)  #  p.Client.QueryRange replaced by  p.Client.Api.QueryRange
	if err != nil {
		log.Warnf("Error found with query %s: %s", query, err)
		continue
}
```

making Query and QueryRange functions not necessary

Reply:
The idea is to, not allow the down-streams to access  the `API`. The downstream should only call the higher level methods like, establishing the client and calling the query methods without direct access to the prom API. And our Prometheus interface should only have access to `API` (because we are having a common implementation here) to run queries and send the results to down-streams, so that the down-streams can only worry calling the parsing the results. For more context on how it can be reused by others: https://github.com/cloud-bulldozer/k8s-netperf/pull/83/files

Reply:
ok, 🤷🏽‍♂️